runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/graph/ProtoOverridesTest.java

```diff
@@ -0,0 +1,253 @@
+            .addAllRootTransformIds(ImmutableList.of("first", "second"))
+            .setComponents(
+                Components.newBuilder()
+                    .putTransforms("first", firstTransform)
+                    .putTransforms(
+                        "second",
+                        PTransform.newBuilder()
+                            .setSpec(FunctionSpec.newBuilder().setUrn("beam:second"))
+                            .build())
+                    .putPcollections(
+                        "intermediatePc",
+                        PCollection.newBuilder().setUniqueName("intermediate").build())
+                    .putCoders(
+                        "coder",
+                        Coder.newBuilder().setSpec(SdkFunctionSpec.getDefaultInstance()).build()))
+            .build();
+
+    PTransform secondReplacement =
+        PTransform.newBuilder()
+            .addSubtransforms("second_sub")
+            .setSpec(
+                FunctionSpec.newBuilder()
+                    .setUrn("beam:second:replacement")
+                    .setPayload(ByteString.copyFrom("foo-bar-baz".getBytes())))
+            .build();
+    WindowingStrategy introducedWS =
+        WindowingStrategy.newBuilder().setAccumulationMode(Enum.ACCUMULATING).build();
+    RunnerApi.Components extraComponents =
+        Components.newBuilder()
+            .putPcollections(
+                "intermediatePc",
+                PCollection.newBuilder().setUniqueName("intermediate_replacement").build())
+            .putWindowingStrategies("new_ws", introducedWS)
+            .putTransforms("second_sub", PTransform.getDefaultInstance())
+            .build();
+
+    Pipeline updated =
+        ProtoOverrides.updateTransform(
+            "beam:second", p, new TestReplacer(secondReplacement, extraComponents));
+    PTransform updatedSecond = updated.getComponents().getTransformsOrThrow("second");
+
+    assertThat(updatedSecond, equalTo(secondReplacement));
+    assertThat(
+        updated.getComponents().getWindowingStrategiesOrThrow("new_ws"), equalTo(introducedWS));
+    assertThat(
+        updated.getComponents().getTransformsOrThrow("second_sub"),
+        equalTo(PTransform.getDefaultInstance()));
+
+    // TODO: This might not be appropriate. Merging in the other direction might force that callers
+    // are well behaved.
+    assertThat(
+        updated.getComponents().getPcollectionsOrThrow("intermediatePc").getUniqueName(),
+        equalTo("intermediate_replacement"));
+
+    // Assert that the untouched components are unchanged.
+    assertThat(updated.getComponents().getTransformsOrThrow("first"), equalTo(firstTransform));
+    assertThat(
+        updated.getComponents().getCodersOrThrow("coder"),
+        equalTo(p.getComponents().getCodersOrThrow("coder")));
+    assertThat(updated.getRootTransformIdsList(), equalTo(p.getRootTransformIdsList()));
+  }
+
+  @Test
+  public void replacesMultiple() {
+    PTransform firstTransform =
```

Comment:
Same as the earlier comment, you can define this inline while creating the pipeline.

Reply:
done